File: LeetCode/20. Valid Parentheses/Solution.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
        
        if len(s)%2 != 0:
            return False
        
        stack = deque([])
        opens = ['(' , "[" , '{']
        matches = {
            '(' : ')' ,
            '[' :']' ,
            '{' : '}'
        }
        for c in s:
            
            if c in opens:
                stack.appendleft(c)
            else:
                if len(stack) == 0:
                    return False
                last_open = stack.popleft()
                if matches[last_open] != c:
                    return False
        
        return len(stack) == 0

File: LeetCode/20. Valid Parentheses/Solution.java
class Solution {
    public boolean isValid(String s) {
        // int len =0;
        // while( len!=s.length() ){
        //     len=s.length();
        //     s=s.replace("[]","");
        //     s=s.replace("()","");
        //     s=s.replace("{}","");
        // }
        // return len==0;
        char[] stack = new char[s.length() + 1];
        int top = 1;
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack[top++] = c;
            } else if (c == ')' && stack[--top] != '(') {
                return false;
            } else if (c == ']' && stack[--top] != '[') {
                return false;
            } else if (c == '}' && stack[--top] != '{') {
                return false;
            }
        }
        return top == 1;        
    }
}
    
